### PR TITLE
ISPN-6680 InterruptedException during PutAll from Hot Rod client prev…

### PR DESCRIPTION
**Pull request by gustavonalle**
…ents failover

https://issues.jboss.org/browse/ISPN-6680
https://issues.jboss.org/browse/ISPN-6692

### CODE REVIEW

**Comment by gustavonalle** [comment]
Adding ISPN-6692 as well

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/OperationStatus.scala [hunk 1]
```diff
@@ -30,6 +30,7 @@ object OperationStatus extends Enumeration {
    val OperationTimedOut = Value(0x86) // todo: test
    val NodeSuspected = Value(0x87)
    val IllegalLifecycleState = Value(0x88)
+   val Interrupted = Value(0x89)
```
Comment (danberindei) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/OperationStatus.scala]: Unless you actually need a different behaviour from `IllegalLifecycleState`, I would prefer to see a single `ServerNotAvailable` operation status covering both. I know sometimes these exceptions can sometimes come from another node, but it should be the core's job to retry them and/or translate them to `SuspectException`.

Comment (gustavonalle) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/OperationStatus.scala]: My initial though was to piggyback on IllegalLifecycleState, but an ```InterruptedException``` could be throw in other situations apart from server going down. For example, a long running job (script?) that gets explicitly cancelled by the user

**Comment by gustavonalle** [comment]
@danberindei thoughts? TBH I was not willing to bind ```InterruptedException``` to a server stopping, do you think it should?

**Comment by danberindei** [comment]
@gustavonalle If the server isn't stoppping, and a long task is being cancelled, should the client retry it? Could the server send back an `IllegalLifecycleState` status if it caught an `InterruptedException` and the server is stopping, and pass the `InterruptedException` unmodified otherwise?

**Comment by gustavonalle** [comment]
> If the server isn't stoppping, and a long task is being cancelled, should the client retry it? 

IMO no, as It could have been explicitly canceled (maybe the script or task was misbehaving).
Note that in this PR it is not retrying the ```ExecuteOperation``` upon ```InterruptedExceptions```

> Could the server send back an IllegalLifecycleState status if it caught an InterruptedException and the server is stopping, and pass the InterruptedException unmodified otherwise?

It makes sense

**Comment by gustavonalle** [comment]
@danberindei I've simplified the PR, without changing Hot Rod version, and wrapping ```InterruptedException``` inside an ```IllegalLifecycleException```. 
Reasons for this is the need to have this backported to several other branches and that we currently don't have "interruptible" operations in the server. We can review the need to expose cancelations to the client later.

**Comment by gustavonalle** [comment]
rebased

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/logging/Log.java [hunk 2]
```diff
@@ -226,8 +226,8 @@
    IllegalStateException queryParameterNotSet(String filterConverterFactoryName);
 
    @LogMessage(level = WARN)
-   @Message(value = "Server not reachable whilst closing iteration '%s'", id = 4061)
-   void ignoringServerUnreachable(String iterationId);
+   @Message(value = "Ignoring error when closing whilst closing iteration '%s'", id = 4061)
```
Comment (danberindei) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/logging/Log.java]: You have to choose between _when_ and _whilst_ :)

Comment (danberindei) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/logging/Log.java]: I've removed _whilst closing_ myself and pushed to master. Trying to rebase on 8.2.x...

**Comment by danberindei** [comment]
@gustavonalle after rebasing to 8.2.x I got a failure in `DistTopologyChangeUnderLoadSingleOwnerTest.testRestartServerWhilePutting` 
```
java.util.concurrent.ExecutionException: org.infinispan.client.hotrod.exceptions.HotRodClientException:Request for messageId=453 returned server error (status=0x85): java.lang.NullPointerException
```
And this exception in the log:
```
19:47:43,569 DEBUG (HotRodServerWorker-119-1:[]) [HotRodExceptionHandler] Exception caught
java.lang.NullPointerException
	at org.infinispan.server.hotrod.ContextHandler.channelActive(ContextHandler.java:210) ~[infinispan-server-hotrod-8.2.3-SNAPSHOT.jar:8.2.3-SNAPSHOT]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:168) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelActive(AbstractChannelHandlerContext.java:154) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.ChannelInboundHandlerAdapter.channelActive(ChannelInboundHandlerAdapter.java:64) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:168) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelActive(AbstractChannelHandlerContext.java:154) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.ChannelInboundHandlerAdapter.channelActive(ChannelInboundHandlerAdapter.java:64) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:168) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelActive(AbstractChannelHandlerContext.java:154) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.ChannelInboundHandlerAdapter.channelActive(ChannelInboundHandlerAdapter.java:64) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at org.infinispan.server.core.transport.StatsChannelHandler.channelActive(StatsChannelHandler.java:45) [infinispan-server-core-8.2.3-SNAPSHOT.jar:8.2.3-SNAPSHOT]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:168) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelActive(AbstractChannelHandlerContext.java:154) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelActive(DefaultChannelPipeline.java:933) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:458) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:370) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:420) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:358) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:280) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137) [netty-all-4.0.36.Final.jar:4.0.36.Final]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_77]
```

Closing this PR, please open a separate one for 8.2.x.

**Comment by gustavonalle** [comment]
@danberindei https://github.com/infinispan/infinispan/pull/4450 for 8.2.x

**Pull request closed without merging**